cpp/src/groupby/sort/group_nth_element.cu

```diff
@@ -0,0 +1,113 @@
+                  rmm::mr::device_memory_resource *mr, cudaStream_t stream) {
+
+  CUDF_EXPECTS(static_cast<size_t>(values.size()) == group_labels.size(),
+               "Size of values column should be same as that of group labels");
+
+  if (num_groups == 0) {
+    return experimental::empty_like(values);
+  }
+
+  auto output = make_numeric_column(
+      data_type{experimental::type_to_id<size_type>()}, num_groups,
+      mask_state::UNALLOCATED, stream);
+  mutable_column_view output_view = output->mutable_view();
+  auto exec = rmm::exec_policy(stream)->on(stream);
+
+  // include nulls (equivalent to pandas nth(dropna=None) but return nulls for n
+  if (_include_nulls == include_nulls::YES || !values.has_nulls()) {
+    // Returns index of nth value.
+    thrust::transform(exec, group_sizes.begin<size_type>(),
+                      group_sizes.end<size_type>(),
+                      group_offsets.begin(), output_view.begin<size_type>(),
+                      [n, out_of_bounds = values.size()] __device__(
+                          auto group_size, auto group_offset) {
+                        bool nth_within_group =
+                            (n < 0) ? group_size >= (-n) : group_size > n;
+                        if (nth_within_group)
+                          return group_offset + ((n < 0) ? group_size + n : n);
+                        else
+                          return out_of_bounds;
+                      });
+  } else { // skip nulls (equivalent to pandas nth(dropna='any'))
+    // Returns index of nth value.
+    thrust::fill(exec, output->mutable_view().begin<size_type>(),
+                 output->mutable_view().end<size_type>(),
+                 values.size()); // for out of bounds
+    auto values_view = column_device_view::create(values);
+    auto bitmask_iterator = thrust::make_transform_iterator(
+        experimental::detail::make_validity_iterator(*values_view),
+        [] __device__(auto b) { return static_cast<size_type>(b); });
+    rmm::device_vector<size_type> intra_group_index(values.size());
+    // intra group index for valids only.
+    thrust::exclusive_scan_by_key(exec, group_labels.begin(),
+                                  group_labels.end(), bitmask_iterator,
+                                  intra_group_index.begin());
+    // gather the valid index == n
+    thrust::scatter_if(
+        exec, thrust::make_counting_iterator<size_type>(0),
+        thrust::make_counting_iterator<size_type>(0) + values.size(),
+        group_labels.begin(),                         // map
+        thrust::make_counting_iterator<size_type>(0), // stencil
+        output->mutable_view().begin<size_type>(),
+        [n, bitmask_iterator,
+         intra_group_index =
+             intra_group_index.begin()] __device__(auto i) -> bool {
+          return (bitmask_iterator[i] && intra_group_index[i] == n);
+        });
+  }
+  bool nullify_out_of_bounds = thrust::transform_reduce(
```

Comment:
This is an expensive operation just to determine if `nullify_out_of_bounds` should be true or false.

Can it just be always set to true?

Reply:
If I set it to `true` always, `detail::gather` creates a null_mask for output column always. That seems unnecessary. (also unit tests `expect_columns_equal` fails).

Reply:
Use `expect_columns_equivalent`. 

You'd have to benchmark what is more expensive. The `transform_reduce` or making the bitmask.

Reply:
```
#nullify_out_of_bounds =  transform_reduce
------------------------------------------------------------------------------
Benchmark                                    Time             CPU   Iterations
------------------------------------------------------------------------------
Groupby/Nth/10000/manual_time            0.222 ms        0.247 ms         3102
Groupby/Nth/100000/manual_time           0.226 ms        0.249 ms         3053
Groupby/Nth/1000000/manual_time          0.317 ms        0.339 ms         2202
Groupby/Nth/10000000/manual_time          1.82 ms         1.84 ms          332
Groupby/Nth/100000000/manual_time          506 ms          506 ms            1

#true
------------------------------------------------------------------------------
Benchmark                                    Time             CPU   Iterations
------------------------------------------------------------------------------
Groupby/Nth/10000/manual_time            0.199 ms        0.223 ms         3475
Groupby/Nth/100000/manual_time           0.202 ms        0.225 ms         3412
Groupby/Nth/1000000/manual_time          0.292 ms        0.315 ms         2385
Groupby/Nth/10000000/manual_time          1.79 ms         1.81 ms          337
Groupby/Nth/100000000/manual_time          506 ms          507 ms            1
```

~10% for small sizes (<1M). not significant  (<1%) for large sizes(>1M values).

Reply:
Is this benchmark run with keys pre-sorted (`sorted::YES`)? Otherwise the cost of sorting might overshadow the performance difference.

Reply:
no. not presorted. Thanks.
```

Comparing old.json to new.json
Benchmark                                             Time             CPU      Time Old      Time New       CPU Old       CPU New
----------------------------------------------------------------------------------------------------------------------------------
Groupby/Basic/10000/manual_time                    -0.1210         -0.1093        223502        196461        247976        220866
Groupby/Basic/100000/manual_time                   -0.1105         -0.1007        227858        202671        250805        225551
Groupby/Basic/1000000/manual_time                  -0.0909         -0.0830        265929        241753        290356        266258
Groupby/Basic/10000000/manual_time                 -0.0413         -0.0397        623140        597375        647968        622226
Groupby/Basic/100000000/manual_time                -0.0049         -0.0049       4272100       4251059       4296759       4275909
```
For small sizes ~10% improvement. For large size, not significant. 
remove `transform_reduce`.

Reply:
Cool! Benchmarks are great.